python/servo/visual_studio.py

```diff
@@ -12,80 +12,141 @@
+                if os.path.exists(installation_path):
+                    yield VisualStudioInstallation(
+                        version_number=version_number,
+                        installation_path=installation_path,
+                        vc_install_path=os.path.join(installation_path, "VC")
+                    )
+
+
+def find_compatible_msvc_with_environment_variables() -> Optional[VisualStudioInstallation]:
+    installation_path = os.environ.get('VSINSTALLDIR')
+    version_number = os.environ.get('VisualStudioVersion')
+    if not installation_path or not version_number:
+        return None
+    vc_install_path = os.environ.get("VCINSTALLDIR", os.path.join(installation_path, "VC"))
+    if not os.path.exists(installation_path) or not os.path.exists(vc_install_path):
+        return None
     return VisualStudioInstallation(
         version_number=version_number,
-        installation_path=vsinstalldir,
+        installation_path=installation_path,
         vc_install_path=vc_install_path,
     )
 
 
-def find_windows_sdk_installation_path(vs_platform: str) -> str:
+def find_msvc_installations() -> VisualStudioInstallation:
+    for installation in find_compatible_msvc_with_vswhere():
+        yield installation
+    for installation in find_compatible_msvc_with_path():
+        yield installation
+    installation = find_compatible_msvc_with_environment_variables()
+    if installation:
+        yield installation
+
+
+def find_msvc_redist_dirs(vs_platform: str) -> Generator[str, None, None]:
+    installations = sorted(set(list(find_msvc_installations())), reverse=True)
+    if not installations:
+        print("Can't find a Visual Studio installation. "
+              "Please set the VSINSTALLDIR and VisualStudioVersion environment variables")
+        sys.exit(1)
+
+    tried = []
+    for installation in installations:
+        redist_dir = os.path.join(installation.vc_install_path, "Redist", "MSVC")
+        if not os.path.isdir(redist_dir):
+            tried.append(redist_dir)
+            continue
+
+        for subdirectory in os.listdir(redist_dir)[::-1]:
+            redist_path = os.path.join(redist_dir, subdirectory)
+            for redist_version in MSVC_REDIST_VERSIONS:
+                # there are two possible paths
+                # `x64\Microsoft.VC*.CRT` or `onecore\x64\Microsoft.VC*.CRT`
+                path1 = os.path.join(vs_platform, "Microsoft.{}.CRT".format(redist_version))
+                path2 = os.path.join("onecore", vs_platform, "Microsoft.{}.CRT".format(redist_version))
+                for path in [path1, path2]:
+                    path = os.path.join(redist_path, path)
+                    if os.path.isdir(path):
+                        yield path
+                    else:
+                        tried.append(path)
+
+    print("Couldn't locate MSVC redistributable directory. Tried:", file=sys.stderr)
+    for path in tried:
+        print(f"  * {path}")
```

Comment:
do we not need to redirect this to `sys.stderr` as well?

Reply:
Yep, I think so. I'll fix that.